### PR TITLE
Don't allow recomputing a node that *must* be materialized in the backwards pass

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89171

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: functorch/_src/partitioners.py [hunk 1]
```diff
@@ -361,7 +362,20 @@ def ban_recomputation(node):
                 return False
             if node.target in [aten.lift_fresh_copy.default, aten.lift_fresh.default]:
                 return False
-            if compiler == "inductor" and node.dist_from_bw > 3:
+
+            # If a node *must* be materialized in the backwards pass, then we
+            # should never recompute it. This is a pretty subtle point.  In
+            # general, the assumption we make is that recomputing a node in the
+            # backwards pass is "free". However, if a node must be materialized
+            # in the backwards pass, then recomputing it is never free.
```
Comment (wconstab) [inline functorch/_src/partitioners.py:370]: i'm still confused by this explanation.

1) what's behind the assumption of "free" recomputation? I would think, heuristcally, (a) lower latency op,  and/or (b) fusible into consumer, probably bw-bound

2) what does it mean that a node 'must' be materialized in bw?  isn't that true for all nodes used in backwards?

3) is the `not is_fusible` clause below really the most important part?  if there isn't a user in bw, then i'd assume we never recompute anyway, why would we

I think i agree with your code but the comment seems misleading.
  * i would not assume it's free to recompute a node in backwards, i would assume it has some cost model
  * recomputed nodes that can be fused into consumers in backwards may actually be "free", right? (or as close as it gets)

Comment (ngimel) [reply functorch/_src/partitioners.py:370]: >  i would assume it has some cost model

There is a cost model, and it is "recomputing is free", it works for the most part, sometimes it fails (duh) e.g. when there's a long chain of math operations and the op as a whole becomes non-bw bound. 
Yes, `not is_fusible` is important, that means that the node must be realized in backward (e.g. as an input to convolution backward or matmul backward), if it's fusible, it exists transiently in registers of the generated kernel, so who cares. 
Do you want to suggest different comment wording?

Comment (wconstab) [reply functorch/_src/partitioners.py:370]: Would this suffice?

"We generally want to ban recomputation of ops used in backwards which are not fusible, since they would incur a cost to recompute."

Comment (Chillee) [reply functorch/_src/partitioners.py:370]: @wconstab Possibly, yeah, but the reason why that's true is a bit subtle imo.

Comment (wconstab) [reply functorch/_src/partitioners.py:370]: lol. eli5?  or 4?

Comment (Chillee) [reply functorch/_src/partitioners.py:370]: @wconstab Or like, I think your explanation covers it?

Basically, the general idea behind this rematerialization optimization is that "computation is cheap, memory accesses are expensive". So, we try to choose the activations that minimizes the total memory accesses.

Broadly speaking, there were 2 sources of memory accesses we were trying to minimize (well... related to the activations).

1. Writes from the forwards pass. 
2. Reads from the backwards pass.

So, writes from the forwards pass are relatively simple to minimize. Each activation you have to save increases your memory accesses. So, the less activations you save, the better. There are some subtleties here. For example, if a tensor *always* needs to be materialized, then choosing to save it doesn't cost you any extra memory accesses.

Reads from the backwards pass are also relatively simple to minimize. Each activation you save, you're going to need to read it from the backwards pass. 

However, one thing we weren't taking into account is that doing recomputation can also increase *writes* in your backwards pass. If you choose to recompute a value that *always* needs to be materialized in the backwards pass, then you might be doing extra *writes* in your backwards pass as well.

So, one thing that's a bit subtle is that I think this is *never* optimal (from a runtime perspective). Recomputing a tensor (that must be materialized in the backwards pass) might decrease writes in your forwards pass, but it results in a corresponding increase in writes in your backwards pass.

NOTE: Crap, as I was writing this I realized that I don't think I'm right. Although recomputing a node that *must* be materialized in your backwards pass won't save you any writes (and so our cost function is wrong), it *can* save you reads.

Imagine in your backwards you had something like
```
def backwards(a: Tensor[10], b: Tensor[10], c: Tensor[10]):
    c_recompute: Tensor[100] = f(a, b)
    d: Tensor[100] = extern_kernel(c)
    e: Tensor[10] = g(c)
```
In this case, it is still potentially profitable to recompute `c` in the backwards pass, as instead of reading from `c` twice, we can read from `c` only once.

Comment (ezyang) [reply functorch/_src/partitioners.py:370]: Can this live as a comment in the code lol

Comment (ngimel) [reply functorch/_src/partitioners.py:370]: @chillee in the convolution case we were talking about, the node (convolution `input`) must be materialized in forward, and it must be materialized in backward, so recomputing it is never profitable. However, if node didn't have to be materialized in forward, then it still might be profitable to recompute it, as in your snippet above.

Comment (wconstab) [reply functorch/_src/partitioners.py:370]:
```
def backwards(a: Tensor[10], b: Tensor[10], c: Tensor[10]):
    c_recompute: Tensor[100] = f(a, b)
    d: Tensor[100] = extern_kernel(c)
    e: Tensor[10] = g(c)
```

not sure im reading this right.  c_recompute is size 100 but c is size 10.  isn't c_recompute just 'c'  but recomputed?  in this example, it seems to me saving c from forward is optimal since that gives 1 write of C and 2 reads during bwd?

Comment (Chillee) [reply functorch/_src/partitioners.py:370]: @wconstab Sorry the example was poorly written.

```
def backwards(a: Tensor[10], b: Tensor[10]):
    c_recompute: Tensor[100] = f(a, b)
    d: Tensor[100] = extern_kernel(c_recompute)
    e: Tensor[10] = g(c_recompute)
```

**Comment by ezyang** [comment]
I think this broke XLNetLMHeadModel

**Comment by desertfire** [comment]
> I think this broke XLNetLMHeadModel

cc @williamwen42
